### PR TITLE
Skip helper when no choice necessary

### PR DESCRIPTION
**Pull request by hickford**
Usability improvement for GitHubAuthentication.

I also wrote some unit tests.

### CODE REVIEW

**Review by Deeprai32037i — APPROVED** [review]

File: src/shared/GitHub/GitHubHostProvider.cs [hunk 1]
```diff
@@ -316,6 +316,11 @@ public static bool IsGitHubDotCom(string targetUrl)
 
         public static bool IsGitHubDotCom(Uri targetUri)
         {
+            if (targetUri is null)
+            {
+                throw new ArgumentNullException(nameof(targetUri));
+            }
```
Comment (mjcheetham) [inline src/shared/GitHub/GitHubHostProvider.cs:322]: We have a set of arg validation helpers that would be better to use here:

```suggestion
            EnsureArgument.NotNull(targetUri, nameof(targetUri));
```

..or for a URI, might even be better to validate for an absolute URI:

```suggestion
            EnsureArgument.AbsoluteUri(targetUri, nameof(targetUri));
```

(this also checks for `null` too)

File: src/shared/GitHub/GitHubAuthentication.cs [hunk 2]
```diff
@@ -79,6 +76,19 @@ public async Task<AuthenticationPromptResult> GetAuthenticationAsync(Uri targetU
...
+            {
+                case AuthenticationModes.Browser:
+                    // no user interaction required
+                    return new AuthenticationPromptResult(AuthenticationModes.Browser);
+                case AuthenticationModes.Device:
+                    // no user interaction required
+                    return new AuthenticationPromptResult(AuthenticationModes.Device);
+                default:
+                    break;
+            }
```
Comment (mjcheetham) [inline src/shared/GitHub/GitHubAuthentication.cs:89]: Let's be consistent with the above `if (modes == .None)` style here:

```suggestion
            // If there is no mode choice to be made and no interaction required,
            // just return that result.
            if (modes == AuthenticationModes.Browser ||
                modes == AuthenticationModes.Device)
            {
                    return new AuthenticationPromptResult(modes);
            }
```

File: Directory.Build.props [hunk 3]
```diff
@@ -23,7 +23,7 @@
   <ItemGroup>
     <!-- All projects should use Nerdbank.GitVersioning for consistent version numbers -->
     <PackageReference Include="Nerdbank.GitVersioning">
-      <Version>3.3.37</Version>
```
Comment (mjcheetham) [inline Directory.Build.props:26]: Please rebase on the latest `main` branch; I've since merged in your PR (#584) to update this package.

**Review by mjcheetham — CHANGES_REQUESTED** [review]
Looks good thanks! Just a few style changes and please rebase to pick-up the new Nerdbank.GitVersioning package commit now in `main`.

File: Directory.Build.props (hunk 3 above)
Comment (hickford) [reply Directory.Build.props]: Done.

**Review by mjcheetham — APPROVED** [review]
Great stuff! Thanks!

**Pull request merged**